doc/source/ray-observability/monitoring-debugging/troubleshoot-failures.rst

```diff
@@ -1,44 +1,252 @@
+  Memory on the node (IP: 10.0.62.231, ID: e5d953ef03e55e26f13973ea1b5a0fd0ecc729cd820bc89e4aa50451) where the task (task ID: 43534ce9375fa8e4cd0d0ec285d9974a6a95897401000000, name=allocate_memory, pid=11362, memory used=1.25GB) was running was 27.71GB / 28.80GB (0.962273), which exceeds the memory usage threshold of 0.95. Ray killed this worker (ID: 6f2ec5c8b0d5f5a66572859faf192d36743536c2e9702ea58084b037) because it was the most recently scheduled task; to see more information about memory usage on this node, use `ray logs raylet.out -ip 10.0.62.231`. To see the logs of the worker, use `ray logs worker-6f2ec5c8b0d5f5a66572859faf192d36743536c2e9702ea58084b037*out -ip 10.0.62.231.`
+  Top 10 memory users:
+  PID	MEM(GB)	COMMAND
+  410728	8.47	510953	7.19	ray::allocate_memory
+  610952	6.15	ray::allocate_memory
+  711164	3.63	ray::allocate_memory
+  811156	3.63	ray::allocate_memory
+  911362	1.25	ray::allocate_memory
+  107230	0.09	python test.py --num-tasks 2011327	0.08	/home/ray/anaconda3/bin/python /home/ray/anaconda3/lib/python3.9/site-packages/ray/dashboard/dashboa...
+
+  Refer to the documentation on how to address the out of memory issue: https://docs.ray.io/en/latest/ray-core/scheduling/ray-oom-prevention.html.
+
+Ray memory monitor also periodically prints the aggregated out-of-memory killer summary to Ray drivers.
+
+.. code-block:: bash
+
+  (raylet) [2023-04-09 07:23:59,445 E 395 395] (raylet) node_manager.cc:3049: 10 Workers (tasks / actors) killed due to memory pressure (OOM), 0 Workers crashed due to other reasons at node (ID: e5d953ef03e55e26f13973ea1b5a0fd0ecc729cd820bc89e4aa50451, IP: 10.0.62.231) over the last time period. To see more information about the Workers killed on this node, use `ray logs raylet.out -ip 10.0.62.231`
+  (raylet) 
+  (raylet) Refer to the documentation on how to address the out of memory issue: https://docs.ray.io/en/latest/ray-core/scheduling/ray-oom-prevention.html. Consider provisioning more memory on this node or reducing task parallelism by requesting more CPUs per task. To adjust the kill threshold, set the environment variable `RAY_memory_usage_threshold` when starting Ray. To disable worker killing, set the environment variable `RAY_memory_monitor_refresh_ms` to zero.
+
+Ray Dashboard's :ref:`metrics page <dash-metrics-view>` and :ref:`event page <dash-event>` also provides the out-of-memory killer-specific events and metrics.
+
+.. image:: ../images/oom-metrics.png
+    :align: center
+
+.. image:: ../images/oom-events.png
+    :align: center
+
+Debug Out-of-Memory Errors
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If tasks or actors are failed by Out-of-memory errors, they are retried based on :ref:`retry policies <ray-oom-retry-policy:>`. 
+However, it is often more desirable to find the root causes of memory issues and fix them instead of relying on fault tolerance mechanisms.
+This section explains how to debug out-of-memory errors in Ray.
+
+**Step 1: Find tasks and actors that use high memory**
```

Comment:
Why are they steps in sequential order actually? It feels to me these are various methods that one could check/verify OOM problems, rather than steps one would follow? (Also, if it's step, I feel it should be more concise and less verbose with all the ray background knowledge)

Reply:
good point. I believe it is difficult to make it concise instruction especially when this doc is not for internal, but external facing (k8s debugging doc is the same https://kubernetes.io/docs/tasks/debug/debug-application/debug-service/). I just made it as an each section rather than steps.